src/software/jetson_nano/services/network.cpp

```diff
@@ -36,28 +33,51 @@ std::tuple<TbotsProto::PrimitiveSet, TbotsProto::World> NetworkService::poll(
 void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
 {
     std::scoped_lock<std::mutex> lock(primitive_set_mutex);
-    primitive_set_msg = input;
-
-    // TODO(#2727): Implement a recent packet loss instead of overall
-    total_primitives_lost += input.sequence_number() - last_primitive_sequence_number - 1;
+    const uint64_t& seq_num = input.sequence_number();
 
-    float packet_loss = static_cast<float>(total_primitives_lost) /
-                        static_cast<float>(input.sequence_number());
+    // If the primitive set seems very out of date, then this is likely due to an AI
+    // reset. Clear the queue
+    if (!recent_primitive_set_seq_nums.empty() &&
+        seq_num + RECENT_PACKET_LOSS_PERIOD <= recent_primitive_set_seq_nums.back())
+    {
+        recent_primitive_set_seq_nums = std::queue<uint64_t>();
+        LOG(WARNING)
+            << "Old primitive set received. Resetting primitive set sequence number tracking";
+    }
+    else if (!recent_primitive_set_seq_nums.empty() &&
+             seq_num <= recent_primitive_set_seq_nums.back())
+    {
+        // If the primitive set is older than the last received primitive set, then ignore
+        // it
+        return;
+    }
+    primitive_set_msg = input;
+    recent_primitive_set_seq_nums.push(seq_num);
 
-    if (packet_loss > PACKET_LOSS_WARNING_THRESHOLD)
+    // Pop sequence numbers of primitive sets that are no longer recent
+    while (seq_num - recent_primitive_set_seq_nums.front() >= RECENT_PACKET_LOSS_PERIOD)
     {
-        LOG(WARNING) << "Primitive packet loss more than "
-                     << PACKET_LOSS_WARNING_THRESHOLD << "% ";
+        recent_primitive_set_seq_nums.pop();
     }
 
-    last_primitive_sequence_number = input.sequence_number();
+    uint64_t expected_primitive_set_count =
+        std::min(seq_num + 1, static_cast<uint64_t>(RECENT_PACKET_LOSS_PERIOD));
```

Comment:
Curious what the +1 is for? Is it to account for the sequence number starting at 0?
Could you add a brief comment describing it.

Reply:
Yeah, the +1 is to account for the sequence number starting at 0. I'll add a comment describing that.